### PR TITLE
[JENKINS-49707] If a KubernetesComputer disconnects, remove the KubernetesSlave

### PR DESCRIPTION
**Pull request by jglick**
Downstream of https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/104. If the pod has disconnected, I am assuming it will not be able to reconnect, and the agent state is lost—true? If so, there is no point in letting any running `sh` step continue: it should fail now.

### CODE REVIEW

**Comment by jglick** [comment]
Seems `RestartPipelineTest` is broken by these changes. Investigating.

**Comment by jglick** [comment]
Hmm, live tests fail because I did not consider this RBAC scenario:

```
WARNING	o.c.j.p.k.pod.retention.Reaper#activate: failed to set up watcher on kubernetes
io.fabric8.kubernetes.client.KubernetesClientException: pods is forbidden: User "system:serviceaccount:…" cannot watch pods at the cluster scope
```

I suppose the answer is to watch only within the `KubernetesCloud.namespace` (also do a startup-time `list pods` in this namespace), and do not attempt to clean up agents associated with pods using a custom namespace, which I assume is rare anyway.

**Comment by jglick** [comment]
With @schottsfired’s help I got 61307bf tested in a semirealistic environment.

**Comment by schottsfired** [comment]
Brilliant @jglick! Looks good on my end.

**Comment by jglick** [comment]
Perhaps related to #201.

**Comment by jglick** [comment]
Tested using [AWS Quickstart for CloudBees Core](https://aws.amazon.com/quickstart/architecture/cloudbees-core/) on EKS with a Pipeline targeted to Spot instances. Simulated a termination in various ways. `kubectl delete pod`:

```
Cannot contact …: hudson.remoting.ChannelClosedException: Channel "unknown": Remote call on JNLP4-connect connection from … failed. The channel is closing down or has closed down
```

`sudo halt -ff`:

```
java.net.SocketTimeoutException: sent ping but didn't receive pong within 1000ms (after 182 successful ping/pongs)
	at okhttp3.internal.ws.RealWebSocket.writePingFrame(RealWebSocket.java:546)
	at okhttp3.internal.ws.RealWebSocket$PingRunnable.run(RealWebSocket.java:530)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Cannot contact …: java.lang.InterruptedException
```

…and in either case the agent does get removed and the build aborted. :tada:

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java [hunk 1]
```diff
@@ -187,6 +191,37 @@ public void runInPodWithRestart() throws Exception {
...
+            // Note that the test is cheating here slightly.
+            // The watch in Reaper is still running across the in-JVM restarts,
+            // whereas in production it would have been cancelled during the shutdown.
+            // But it does not matter since we are waiting for the agent to come back online after the restart,
+            // which is sufficient trigger to reactivate the reaper.
+            // Indeed we get two Reaper instances running, which independently remove the node.
+            deletePods(cloud.connect(), getLabels(this, name), false);
+            r.assertBuildStatus(Result.ABORTED, r.waitForCompletion(b));
+            while (!JenkinsRule.getLog(b).contains(new ExecutorStepExecution.RemovedNodeCause().getShortDescription())) {
+                // TODO JenkinsRule.waitForMessage has a race condition w.r.t. the termination cause printed by WorkflowRun.finish
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java]: Probably same as what is mentioned in [JENKINS-46076](https://issues.jenkins-ci.org/browse/JENKINS-46076); have not bothered to deal with it yet.

Comment (jglick) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java]: see #496

File: Jenkinsfile [hunk 2]
```diff
@@ -1 +1,4 @@
-buildPlugin(configurations: buildPlugin.recommendedConfigurations().findAll { it.platform == 'linux' })
+buildPlugin(configurations: [
```
Comment (jglick) [inline Jenkinsfile]: Without write access, the PR builds are going to fail in `InjectedTest`.

File: pom.xml [hunk 3]
```diff
@@ -144,19 +146,19 @@
...
       <artifactId>workflow-support</artifactId>
-      <version>3.3</version>
+      <version>${workflow-support-plugin.version}</version>
       <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.jenkins-ci.plugins.workflow</groupId>
       <artifactId>workflow-durable-task-step</artifactId>
-      <version>2.28</version>
+      <version>2.32-rc934.23119201f0b5</version> <!-- TODO https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/104 -->
```
Comment (dwnusbaum) [inline pom.xml]:
```suggestion
      <version>2.32</version>
```

**Pull request merged**